skbio/tree/_gme.py

```diff
@@ -0,0 +1,452 @@
+    """Return the average distance between the leaves of two subtrees.
+
+    Distances between nodes are calculated using a distance matrix.
+
+    """
+    nodelist1 = _tip_or_root(node1)
+    nodelist2 = _tip_or_root(node2)
+    df = dm.between(nodelist1, nodelist2)
+    return df["value"].mean()
+
+
+def _tip_or_root(node):
+    """Get name(s) of a node if it's a tip or root, otherwise its descending tips."""
+    if node.is_tip() or node.is_root():
+        return [node.name]
+    else:
+        return [x.name for x in node.tips()]
+
+
+def _average_distance_upper(node1, node2, dm):
+    """Return the average distance between the leaves of two subtrees.
+
+    Used for subtrees which have a set of tips that are the complement
+    of the set of tips that are descendants from the node defining
+    the subtree.
+
+    Given an internal edge of a binary tree, exactly one adjacent edge
+    will connect to a node defining a subtree of this form.
+
+    """
+    nodelist1 = _tip_or_root(node1)
+    if node2.is_root():
+        nodelist2 = []
+    # Second subtree serves as the tree with a set of tips
+    # complementary to the set of tips that descend from the
+    # corresponding second node.
+    else:
+        root2 = node2.root()
+        nodelist2 = [root2.name]
+        nodelist2.extend(root2.subset() - node2.subset())
+    df = dm.between(nodelist1, nodelist2)
+    return df["value"].mean()
+
+
+def _subtree_count(subtree):
+    """Return the number of leaves in a subtree.
+
+    Assumes the root as a leaf node.
+
+    """
+    if subtree.is_tip() or subtree.is_root():
+        return 1
+    else:
+        return subtree.count(tips=True)
+
+
+def _average_subtree_distance(a, b, a1, a2, dm):
+    """Return the average distance between two subtrees."""
+    return (
+        _subtree_count(a1) * _average_distance(a1, b, dm)
+        + _subtree_count(a2) * _average_distance(a2, b, dm)
+    ) / _subtree_count(a)
+
+
+def _average_distance_matrix(tree, dm):
+    """Return the matrix of distances between pairs of subtrees."""
+    ordered = list(tree.postorder(include_self=False))
+    n = len(ordered)
+    r = tree.root()
+    taxa_size = r.count(tips=True) + 1
+    adm = np.empty((n, n))
+    for i, a in enumerate(ordered):
+        # skip over unique descendant
+        if a in tree.children:
+            continue
+        # find the average distance between given node and root
+        if a.is_tip():
+            adm[n - 1, i] = adm[i, n - 1] = dm[a.name, r.name]
+        else:
+            a1, a2 = a.children
+            adm[n - 1, i] = adm[i, n - 1] = _average_subtree_distance(a, r, a1, a2, dm)
+        # find the average distance between first node and a second node
+        # which is above the first node in the postorder as well as an ancestor
+        for j in range(i + 1, n - 1):  # part (a)
+            b = ordered[j]
+            # skipping over ancestors
+            if b in a.ancestors():
```

Comment:
This `a` remains the same throughout the `for` loop. However, calling `ancestors` costs time in each iteration. A faster way would be calling it only once prior to the loop and reuse the result in each iteration.